packages/ui/Scrollable/Scrollable.js

```diff
@@ -688,95 +698,81 @@ class ScrollableBase extends Component {
 			overDistance = -position;
 		} else if (position > maxPos) {
 			overDistance = position - maxPos;
+		} else {
+			return 0;
 		}
 
-		return clamp(0, 1, 2 * overDistance / baseSize);
+		return Math.min(1, 2 * overDistance / baseSize);
 	}
 
-	updateOverscrollEffectOnPosition = (targetX, targetY, type, ratio) => {
-		this.updateOverscrollEffect('horizontal', targetX, type, ratio);
-		this.updateOverscrollEffect('vertical', targetY, type, ratio);
+	applyOverscrollEffect = (orientation, edge, type, ratio) => {
+		this.props.applyOverscrollEffect(orientation, edge, type, ratio);
+		this.setOverscrollStatus(orientation, edge, type === overscrollTypeOnce ? overscrollTypeDone : type, ratio);
 	}
 
-	updateOverscrollEffectByDirection = (orientation, position, type, ratio = null) => {
-		const applyOverscrollEffect = this.props.applyOverscrollEffect;
+	checkAndApplyOverscrollEffect = (orientation, edge, type, ratio) => {
+		const
+			isVertical = (orientation === 'vertical'),
+			curPos = isVertical ? this.scrollTop : this.scrollLeft,
+			maxPos = this.getScrollBounds()[isVertical ? 'maxTop' : 'maxLeft'];
 
-		if (applyOverscrollEffect && this.getScrollabilities(orientation)) {
-			const
-				isVertical = (orientation === 'vertical'),
-				curPos = isVertical ? this.scrollTop : this.scrollLeft,
-				maxPos = this.getScrollBounds()[isVertical ? 'maxTop' : 'maxLeft'];
-			let
-				newType = type,
-				newRatio = (ratio === null) ? 1 : ratio;
-
-			if (newRatio > 0) {
-				if (type === overscrollTypes.scrolling) {
-					if (position === 'before' && curPos <= 0) { // Already on the beginning egde
-						applyOverscrollEffect(orientation, 'before', type, ratio);
-						newType = overscrollTypes.none;
-						newRatio = 0;
-					} else if (position >= 'after' && curPos >= maxPos) { // Already on the ending egde
-						applyOverscrollEffect(orientation, 'after', type, ratio);
-						newType = overscrollTypes.none;
-						newRatio = 0;
-					}
-				} else if (type === overscrollTypes.sticking) {
-					if (position === 'before' && curPos <= 0) { // Already on the beginning egde
-						applyOverscrollEffect(orientation, 'before', type, ratio);
-					} else if (position === 'after' && curPos >= maxPos) { // Already on the ending egde
-						applyOverscrollEffect(orientation, 'after', type, ratio);
-					}
-				}
+		if (edge === 'before' && curPos <= 0) { // On the beginning edge
+			this.applyOverscrollEffect(orientation, 'before', type, ratio);
+		} else if (edge === 'after' && curPos >= maxPos) { // On the ending edge
+			this.applyOverscrollEffect(orientation, 'after', type, ratio);
+		} else {
+			this.setOverscrollStatus(orientation, edge, type, ratio);
+		}
+	}
 
-				this.setOverscrollStatus(orientation, newType, newRatio);
-			}
+	clearOverscrollEffect = (orientation, edge) => {
+		if (this.props.clearOverscrollEffect) {
+			this.props.clearOverscrollEffect(orientation, edge);
```

Comment:
I think we can call this function only if the type is not 'none' from UI side not in moonstone.